compiler/x/codegen/DataSnippet.cpp

```diff
@@ -48,19 +49,29 @@ void
 TR::IA32DataSnippet::addMetaDataForCodeAddress(uint8_t *cursor)
    {
    // add dummy class unload/redefinition assumption.
+   bool needRelocation = cg()->comp()->compileRelocatableCode() || TR::Compiler->cls.classUnloadAssumptionNeedsRelocation();
+   if (needRelocation)
+      {
+      cg()->addExternalRelocation(new (TR::comp()->trHeapMemory())
+                               TR::ExternalRelocation(cursor, nullptr, TR_ClassUnloadAssumption, cg()),
+                               __FILE__, __LINE__, self()->getNode());
+      }
+
    if (_isClassAddress)
       {
       if (TR::Compiler->target.is64Bit())
          {
-         cg()->jitAddPicToPatchOnClassUnload((void*)-1, (void *) cursor);
+         if (!needRelocation)
+            cg()->jitAddPicToPatchOnClassUnload((void*)-1, (void *) cursor);
```

Comment:
I don't know if this will work as intended. IIRC it's possible to AOT-compile a method at runtime and then run it without processing the relocations. If that's still possible then you will need to register the assumption because the relocation may not be processed. Perhaps @dsouzai knows if that's still possible.

Reply:
Even on the AOT compile run we go through and process the relocations; in fact I think some code depends on the relocations being processed because, for example, when a `lea` gets converted to a `mov + lea`, the immediate value of the `mov` will be garbage (even on the AOT compile run) until it gets relocated.